src/passes/TypeSSA.cpp

```diff
@@ -52,13 +52,113 @@
 #include "ir/names.h"
 #include "ir/possible-constant.h"
 #include "pass.h"
+#include "support/hash.h"
 #include "wasm-builder.h"
 #include "wasm.h"
 
 namespace wasm {
 
 namespace {
 
+// Given some TypeBuilder items that we want to build new types with, this
+// function builds the types in a new rec group.
+//
+// This is almost the same as just calling build(), but there is a risk of a
+// collision with an existing rec group. This function handles that by finding a
+// way to ensure that the new types are in fact in a new rec group.
+//
+// TODO: Move this outside if we find more uses.
+struct BuilderItem {
+  std::variant<Struct, Array> contents;
+  HeapType super;
+};
+std::vector<HeapType>
+makeNewTypesInNewRecGroup(std::vector<BuilderItem>& builderItems,
+                          Module& wasm) {
+  auto num = builderItems.size();
+
+  // Find all the heap types present before we create the new ones. The new
+  // types must not appear in |existingSet|.
+  std::vector<HeapType> existing = ModuleUtils::collectHeapTypes(wasm);
+  std::unordered_set<HeapType> existingSet(existing.begin(), existing.end());
+
+  // Given a builder, set it up with the builder items we were given.
+  auto fillBuilder = [&](TypeBuilder& builder) {
+    for (Index i = 0; i < builderItems.size(); i++) {
+      auto& item = builderItems[i];
+      if (auto* struct_ = std::get_if<Struct>(&item.contents)) {
+        builder[i] = *struct_;
+      } else if (auto* array_ = std::get_if<Array>(&item.contents)) {
+        builder[i] = *array_;
+      } else {
+        WASM_UNREACHABLE("bad variant");
+      }
+      builder[i].subTypeOf(item.super);
+    }
+  };
+
+  // Create a copy of the builder and see if just building happens to work.
+  TypeBuilder builder(num);
+  fillBuilder(builder);
+  builder.createRecGroup(0, num);
+  auto result = builder.build();
+  assert(!result.getError());
+  auto newTypes = *result;
+  assert(newTypes.size() == num);
+
+  // Check for a collision with an existing rec group. Note that it is enough to
+  // check one of the types: either the entire rec group gets merged, so they
+  // are all merged, or not.
+  if (existingSet.count(newTypes[0])) {
+    // Unfortunately there is a conflict. Handle it by adding a "hash" - a
+    // "random" extra item in the rec group that is so outlandish it will
+    // surely (?) never collide with anything. We must loop will doing so, until
+    // we find a hash that does not collide.
+    auto hashSize = num + 10;
+    size_t random = num;
+    while (1) {
```

Comment:
Could this function be made shorter using a do-while loop?

Reply:
I think it's moot now after the refactoring to use a rec group, but even before I'm not sure it would help since the sizes of the builders are different and there is the special last item, so it wasn't the same command twice.